File: Leetcode_Python/insertion_sort_list.py
'''Version 1'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def insertionSortList(self, head):
        if not head or head.next == None:  
            return head
        dummy = ListNode(-1)
        dummy.next = head
        p = head
        while p.next:
            if p.val < p.next.val:
                p = p.next
            else:
                q = dummy
                while q.next.val < p.next.val:
                    q = q.next
                tem = q.next
                q.next = p.next
                p.next = p.next.next
                q.next.next = tem
        return dummy.next
        
'''Version 2'''
class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def insertionSortList(self, head):
        if not head or head.next == None:  
            return head
        
        lst = []
        while head:
            lst.append(head.val)
            head = head.next
        
        for i in xrange(1, len(lst)):
            temp = lst[i]
            j = i - 1
            while j >= 0 and temp < lst[j]:
                lst[j + 1] = lst[j]
                j -= 1
            lst[j + 1] = temp
        
        res = [ListNode(i) for i in lst]
        for i in xrange(len(lst) - 1):
            res[i].next = res[i + 1]
        res[len(lst) - 1].next = None
        return res[0]

File: Leetcode_Java/InsertionSortList.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        this.val = x;
    }
}

public class InsertionSortList {
    public ListNode insertionSortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode pre = dummy;
        ListNode cur = head;
        while (cur.next != null) {
            while (cur.next != null && cur.val <= cur.next.val) {
                cur = cur.next;    
            }
            while (cur.next != null && cur.val > cur.next.val) {
                ListNode temp = cur.next;
                cur.next = cur.next.next;
                while (pre.next.val < temp.val) {
                    pre = pre.next;
                }
                temp.next = pre.next;
                pre.next = temp;
                pre = dummy;
            }
        }
        return dummy.next;
    }

    public static void main(String[] args) {
        InsertionSortList test = new InsertionSortList();
        ListNode[] list = new ListNode[7];
        int[] values = {3, 1, 2, 6, 0, 4, 5};
        for (int i = 0; i < 7; i++) {
            list[i] = new ListNode(values[i]);
            if (i != 0) {
                list[i - 1].next = list[i];
            }
        }
        ListNode head = test.insertionSortList(list[0]);
        while (head != null) {
            System.out.print(head.val + ", ");
            head = head.next;
        }
    }
}